File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        // find the middle 
        ListNode slow = head; 
        ListNode fast = head; 

        while(fast!= null && fast.next != null){
            slow = slow.next; 
            fast = fast.next.next; 
        }
        // reverse after middle 
        ListNode prev = null; 
        ListNode curr = slow.next; 
        slow.next = null;

        while( curr!= null){
            ListNode next = curr.next; 
            curr.next = prev; 
            prev = curr; 
            curr = next;
        }

        // reorder 
        ListNode first = head; 
        ListNode second = prev;  //since prev indicates our head of reversed linked list 

        while( second!= null){
            ListNode temp1 = first.next; 
            ListNode temp2 = second.next; 

            first.next = second; 
            second.next = temp1; 

            first = temp1; 
            second = temp2;
        }
        
    }
}

File: 0143-reorder-list/0143-reorder-list.cpp
class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head || !head->next) return;

        // Step 1: Find the middle node using slow and fast pointers
        ListNode* slow = head;
        ListNode* fast = head;
        
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // Step 2: Reverse second half
        ListNode* prev = nullptr;
        ListNode* curr = slow->next;
        slow->next = nullptr; // Disconnect first and second half

        while (curr) {
            ListNode* next_temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next_temp;
        }

        // Step 3: Merge two halves
        ListNode* first = head;
        ListNode* second = prev;

        while (second) {
            ListNode* tmp1 = first->next;
            ListNode* tmp2 = second->next;

            first->next = second;
            second->next = tmp1;

            first = tmp1;
            second = tmp2;
        }
    }
};
